### DISCUSSION TITLE
请问，可以添加自己的tool script么

### DISCUSSION THREAD

**Discussion by PumpkinXD** [xmake-io/xmake#4640] [Q&A] [answered]
（我不太会表达想法，可能不太通顺/好懂）
在[4548](https://github.com/xmake-io/xmake/discussions/4548)我问过类似的问题，当时是让`xmake`将`cosmocc`认作`gcc`来处理

现在我想让`xmake`将`cosmocc`认作不同的编译器, 但仅仅去掉`gcc@`不搞其他的会报`error: cannot find known tool script for cosmocc`

不知可不可以将内置的gcc的tool script拿出来修改几处后（不清楚[这个](https://github.com/xmake-io/xmake/blob/master/xmake/modules/core/tools/gcc.lua)是不是内置的gcc tool script)

然后放到项目下的某个文件夹下(`./tools`?)是否可行

**Comment by waruqi** [comment, marked as answer]
目前暂时不支持完全自定义未知的 tool，要么走 toolchain + rule ，要么走 gcc@ 走已知的 gcc 逻辑，要么你干脆提 pr ，让 xmake 内置支持 cosmocc

**Reply by axxdzkl** [reply]
商用的IAR，使用iccarm.exe编译，ilink.exe链接，之前使用cmake是原生支持的，转换到xmake是不是需要参照cmake实现自定义tool，怎么使用toolchain+rule支持自定义tool呢？

**Reply by axxdzkl** [reply]
是不是类似这样写load.lua和xmake.lua设置编译器

**Reply by axxdzkl** [reply]
![QQ20240816-235735](https://github.com/user-attachments/assets/1f73e583-adee-4a67-b178-58bdd7be8c94)

**Reply by waruqi** [reply]
如果iccarm 是 gcc 系的可以这么搞，如果是完全未知的编译器，只能提 pr 内置支持
